runtime/metricsview/executor_rewrite_approx_comparisons.go

```diff
@@ -92,6 +97,23 @@ func (e *Executor) rewriteApproxComparisonNode(a *AST, n *SelectNode) bool {
 		n.FromSelect.OrderBy = order
 		n.FromSelect.Limit = a.Root.Limit
 		n.FromSelect.Offset = a.Root.Offset
+
+		// ---- CTE Optimization ---- //
+		// make FromSelect a CTE and set IsCTE flag on FromSelect
+		cte := n.FromSelect
+		n.CTEs = append(n.CTEs, cte)
+		n.FromSelect.IsCTE = true
+
+		// now change the JoinComparisonSelect WHERE clause to use selected dim values from CTE
+		for _, dim := range cte.DimFields {
+			var dimExpr string
+			if dim.Expr != "" {
+				dimExpr = dim.Expr
+			} else {
+				dimExpr = dim.Name
+			}
+			n.JoinComparisonSelect.Where = n.JoinComparisonSelect.Where.and(fmt.Sprintf("%[1]s IS NULL OR %[1]s IN (SELECT %[2]q.%[3]q FROM %[2]q)", dimExpr, cte.Alias, dim.Name), nil)
+		}
```

Comment:
Might be worth constructing an `*Expression` and using `sqlForExpression` to build the SQL here. There are two things to think about:
- Correct SQL escaping of the names
- Correct handling if the dim is unnested (not completely sure if it's needed here or not)

https://github.com/rilldata/rill/blob/4edcc574cb32a6201597069e1e51db96604b67ef/runtime/metricsview/astexpr.go#L10-L13

Reply:
you mean create an expression for `"%[1]s IS NULL OR %[1]s IN (SELECT %[2]q.%[3]q FROM %[2]q)"` first in the same place then use `sqlForExpression` to get expr string instead of `fmt.Sprintf`. Not sure what difference will it make as at this point SelectNode contains ExprNode for where clause which is already deserialized form of the expression, also the dims are also deserialized into FieldNode

Reply:
Just for context, an expression filtering on an unnested dimension which isn't in `DimFields` requires some extra handling, such as this: https://github.com/rilldata/rill/blob/46bf4dec7b6a69d95a6fe24d1af2f00681ec4655/runtime/metricsview/astexpr.go#L238

However, in this case you're only targeting dimensions in `DimFields`, so you're right that that concern does not apply.

However, the concern of escaping still applies. You can fix it here by incorporating the right escape functions (`dialect.EscapeIdentifier` for the dim name and parentheses around the dim expressions).

Reply:
I understand what you mean but my only concern is `Expression` is created at an upper level in `NewAST` before optimizations so its a little odd to go back to that format here.

Reply:
I tried using expression but there its not fully supported yet for this use case
1. `IS` operator is not supported.
2. Writing `NULL` value is not supported.
3. `subquery` in the Expression does not support writing plain (non nested) select stmt referring any table or a CTE.

For now, I have changed to escape dim names and parenthesize dim expr, adding support for 3 will require some thought.